### PR TITLE
L2 driver for KWP6227 (Keyword D3 B0) protocol

### PR DESCRIPTION
**Pull request by aaeegg**

### CODE REVIEW

File: scantool/diag_l0_elm.c [hunk 1]
```diff
@@ -57,6 +57,9 @@ struct elm_device {
 
 	struct diag_serial_settings serial;
 	ttyp *tty_int;			/** handle for tty stuff */
+
+	uint8_t kb1, kb2;	// key bytes from 5 baud init
+	uint8_t atsh[3];	// current header setting for ISO9141
```
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: this should be initialized explicitly somewhere even if it will probably work fine. The code that checks if headers were changed is otherwise reading an uninitialized value on first execution

File: scantool/diag_l0_elm.c [hunk 2]
```diff
@@ -743,13 +752,77 @@ elm_initbus(struct diag_l0_device *dl0d, struct diag_l1_initbus_args *in)
 					break;
 			}
 
+			//set init address
+			if ((dev->elmflags & ELM_323_BASIC) && (in->addr != 0x33)) {
+				fprintf(stderr, FLFMT "elm_initbus: ELM323 doesn't support target address %02X", FL, in->addr);
+				return diag_iseterr(DIAG_ERR_GENERAL);
+			} else {
+				uint8_t iia[15];
+				sprintf((char *) iia, "ATIIA %02X\x0D", in->addr);
```
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: I confess I didn't check, is ATIIA well supported by clones ?

File: scantool/diag_l0_elm.c (hunk 1 above)
Comment (aaeegg) [inline scantool/diag_l0_elm.c]: atsh[3] is initialized to {0,0,0} by the diag_calloc call in elm_new().

File: scantool/diag_l0_elm.c [hunk 3]
```diff
@@ -743,13 +752,77 @@ elm_initbus(struct diag_l0_device *dl0d, struct diag_l1_initbus_args *in)
...
+			if (dev->elmflags & ELM_INITDONE) {
+				uint8_t rxbuf[ELM_BUFSIZE];
+				unsigned int kb1, kb2;
+
+				buf=(uint8_t *) "ATKW\x0D";
+				rv=elm_sendcmd(dl0d, buf, 5, 500, rxbuf);
+				if (rv < 0) {
+					fprintf(stderr, FLFMT "elm_initbus: ATKW failed, continuing anyway\n", FL);
+					// TODO: for clones without ATKW, can
+					// fall back to ATBD. If ATBD response
```
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: Great idea, it could even deal with possible variations in ATBD's buffer layout. TODO, someday P)

File: scantool/diag_l0_elm.c [hunk 4]
```diff
@@ -743,13 +752,77 @@ elm_initbus(struct diag_l0_device *dl0d, struct diag_l1_initbus_args *in)
...
+					dev->kb2 = kb2;
+					in->kb1 = kb1;
+					in->kb2 = kb2;
+				} else {
+					fprintf(stderr, FLFMT "elm_initbus: ATKW failed, continuing anyway\n", FL);
+					rv = 0;
+				}
+			}
+
+			//for KWP6227, set appropriate wakeup message
```
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: hmm, I know this is extra work, but I'd really prefer having this as another IOCTL. It could be quite simple and only implemented on ELM for now, and would move the ugliness of hardcoded constants out of L0 code and into the L2 code.

File: scantool/diag_l0_elm.c [hunk 5]
```diff
@@ -824,6 +897,9 @@ elm_send(struct diag_l0_device *dl0d,
 	if (len <= 0)
 		return diag_iseterr(DIAG_ERR_BADLEN);
 
+	if ((dev->protocol & DIAG_L1_ISO9141) && len <= 3)
```
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: Interesting; I'm not opposed to this but I wonder if you had problems related to this ?

File: scantool/diag_l0_elm.c [hunk 6]
```diff
@@ -834,6 +910,35 @@ elm_send(struct diag_l0_device *dl0d,
 		fprintf(stderr, FLFMT "ELM: sending %d bytes\n", FL, (int) len);
 	}
 
+	if ((dev->protocol & DIAG_L1_ISO9141) && memcmp(dev->atsh, data, 3)) {
```
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: This is what I was referring to earlier,  regarding atsh[3]; on the first execution this is fragile because atsh is not initialized.

File: scantool/diag_l2_kwp6227.c [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+	rv = dl2p_6227_recv(d_l2_conn, 1000, dl2p_6227_request_callback, &rmsg);
+	if (rv < 0) {
+		*errval = rv;
+		return rmsg;
+	}
+	if (rmsg == NULL)
+		*errval = DIAG_ERR_NOMEM;
+
+	return rmsg;
```
Comment (fenugrec) [inline scantool/diag_l2_kwp6227.c]: to be totally rigorous, the normal, error-free exit with "return rmsg" should have an explicit " *errval = 0;"  ; I've had this trigger some compiler warnings in some conditions.

File: scantool/diag_l0_elm.c (hunk 2 above)
Comment (aaeegg) [inline scantool/diag_l0_elm.c]: I believe that most "ELM v1.5" clones support ATIIA. Some clones don't support ATIIA. Depending on the clone, they either respond with "?" or treat ATIIA as ATI.

The ELM's default target address for ISO9141 is 0x33, which is used for J1979 over ISO9141. The new ATIIA code treats a failure as soft if the address is 0x33, so it should still work for J1979 on clones that don't support ATIIA. However, I don't have one of those clones to test with.

KWP6227 isn't possible on clones without ATIIA.

File: scantool/diag_l2_kwp6227.c [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+		free(pX->diag_l2_proto_data);
+		pX->diag_l2_proto_data=NULL;
+	}
+
+	return 0;
+}
+
+static void
+dl2p_6227_timeout(struct diag_l2_conn *d_l2_conn)
+{
```
Comment (fenugrec) [inline scantool/diag_l2_kwp6227.c]: In the current conditions, this is never called (L0_elm sets the L1_DOESKEEPALIVE flag) ! But it can stay there , in case this L2 driver is modified to also accept L0_dumb interfaces.

File: scantool/diag_l0_elm.c (hunk 1 above)
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: OK

File: scantool/diag_l0_elm.c (hunk 2 above)
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: OK. Could you add an " if (in->addr != 0x33) " to only call ATIIA if necessary; the extra delay makes a difference when using the scan / probe commands.

Comment (aaeegg) [inline scantool/diag_l0_elm.c]: OK, I'll do that this weekend.

File: scantool/diag_l0_elm.c (hunk 4 above)
Comment (aaeegg) [inline scantool/diag_l0_elm.c]: I agree. However, there is currently no way to just add an IOCTL in an L0 driver. I would have to add a new _set_keepalive method to struct diag_l0, add corresponding code to diag_l2_ioctl (!) to invoke it, and add stubs to the other L0 drivers. Do you want me to do that?

File: scantool/diag_l0_elm.c (hunk 5 above)
Comment (aaeegg) [inline scantool/diag_l0_elm.c]: Previously, elm_send() required at least 1 payload byte. Now, in ISO9141, it gets the header from L2, so it requires 3 header bytes + 1 payload byte. This change preserves the previous requirement of at least 1 payload byte.

File: scantool/diag_l2_kwp6227.c (hunk 7 above)
Comment (aaeegg) [inline scantool/diag_l2_kwp6227.c]: OK, I'll change this.

File: scantool/diag_l0_elm.c (hunk 4 above)
Comment (fenugrec) [inline scantool/diag_l0_elm.c]: ah, hm, right again. I was supposed to add L0 ioctl handlers (to replace setspeed, initbus, etc)... not a huge job, but doesn't belong in this PR. You can leave that in there, and we can revisit once I've overhauled the L0 ioctl code.

**Comment by fenugrec** [comment]
merged, thanks !
